### ISSUE TITLE
Moving window to upper left corner fails (indent still remaining - both on top and on the left)

### ISSUE THREAD

**Issue by mikkovedru** [linuxmint/cinnamon#9233]
```
* Linux Mint 19.2 Cinnamon (4.2.4)
* 64 bit
* Kernel 5.3.0-42-generic
* NVIDIA Corporation GK107 [GeForce GTX 650]
* Driver: nvidia-driver-435 (435.21-0ubuntu0.18.04.2)
 ```

**Issue**
Moving window to upper left corner fails (indent still remaining - both on top and on the left). This bug has been there for a long time, can't really recall for how long (I have been using Cinnamon from 2014).

**Steps to reproduce**
## Prerequisites

1. System Settings → Keyboard → Shortcuts → Windows → Positioning:
- Move window to left edge: Super+Q
- Move window to right edge: Super+W
- Move window to top edge: Super+A
![2020-04-04_Screenshot_1](https://user-images.githubusercontent.com/2521942/78408787-b23b3380-7610-11ea-891c-6a679bcab85b.png)

2. System Settings → Extensions → Download and enable "gTile". This is an extention that allows to easily change location of the window(s). After enabling:
- Assign global hotkey for gTile as Super+S.
- Enable "3x2" mode (2 rows of 3 columns each) as default.
![2020-04-04_Screenshot_2](https://user-images.githubusercontent.com/2521942/78408796-b8c9ab00-7610-11ea-9aa6-122499a7cfca.png)
gTile extension's settings:
[gtile-settings-json.txt](https://github.com/linuxmint/cinnamon/files/4430118/gtile-settings-json.txt)

## How to reproduce
1. Open Firefox window.
2. Press `Super+S` to open gTile extension.
3. Press `Right` 3 times.
4. Hold `Shift` and press `Down`.
5. Stop pressing `Shift`.
6. Press `Enter`. At this point the window should have moved to occupy both rows of the 3rd column.
7. Press Super+S to open gTile extension.
8. Press `Right`.
9. Hold `Shift` and press `Right` and `Down`.
10. Press `Enter`. At this point the window should have moved to occupy both rows of the 1st and 2nd columns. Except that you can see that there is space on top and on the left. This is a bug.

## Argumentation why is this not a gTile bug?
1. Try to click (and hold) the mouse on the title and move the window higher. You will notice that it acts as if it's already on the very top.

2. Try to hover the Minimize/Maximize/Close buttons or the edges. You will see that the mouse cursor changes not at the appropriate location, but with the offset of a size that is missing from left and top. Clearly an issue with Cinnamon.

3. Pressing `Super+A` to put the window to top edge doesn't do anything. It would have, if Cinnamon would not have this specific bug. Clearly an issue with Cinnamon.

4. Pressing `Super+Q` to put the window to left edge doesn't do anything. It would have, if Cinnamon would not have this specific bug. Clearly an issue with Cinnamon.

5. Pressing `Super+W` (even without previous pressing of Super+A` or Super+Q`) to put the window to right edge works without any problems. The combination moves the window to the right, but also moves it to the top edge, meaning that from the system's point of view the window was on the top edge all along! Pressing `Super+Q` to put the window to left edge works fine after this detour!

## Other information and considerations
This bug doesn't seem to be general (working with all programs, all the time).

This bug seems to affect different programs in different ways. Another strong possibility, that it has a time element to it (for example two windows of the same program are behaving consistently, but differently from each other).

When I have been investigating this bug more closely while writing this bug report, this is what I noticed:
1. Firefox window only bugs when moving from the rightest column to the leftest column (or 2 leftest ones).
2. Doesn't seem to bug with the LibreOffice Writer window. At least not now.
3. gEdit bugs immediately, regardless of which column I put it in (including the middle column). Interestingly, the offset in the left column is clearly bigger than in the central or right column.
4. Nemo file manager window. Some windows bug, some don't.
5. Evince document viewer ( https://wiki.gnome.org/Apps/Evince ) is bugged, but interestingly two different .pdf documents open in the same program are bugging in all kinds of different ways.
6. Strange, new Nemo windows are not bugging anymore...
7. Oh, I closed some "bugging windows/programs", opened a new LibreOffice Writer document and now it bugs as well.

## Conclusion/speculations
1. Maybe bug is treating different programs differently?
2. But at least it seems to be certain that it matter how many programs are being run, in which order they were started, etc.
3. There might be a limit to how many windows can be bugged at the same time?
4. The bug is stackable and is influenced by its own wrong behaviour. Look at the following screenshot, in which I manage to "bug" (organize) 3 or 4 different windows in such a way, that they are counting offset from each other instead of the top left corner.
![2020-04-04_Screenshot_3](https://user-images.githubusercontent.com/2521942/78409906-9e450100-7613-11ea-99e7-3c0516d5bf85.png)

## Other
This bug may or may not have relation with the ignored bug #8546 that has been part of Cinnamon for many years now.

**Comment by leigh123linux** [comment]
So you expect apps that have a minimum size to fit into the 3x2  tile setting?, what screen resolution are you using?

**Comment by mikkovedru** [comment]
> So you expect apps that have a minimum size to fit into the 3x2 tile setting?, what screen resolution are you using?

I don't know what you mean by "minimum size". There is a certain (Firefox, Nemo, LibreOffice Writer, etc) window. Its size can be changed. gTile allows me to easily change both:
1. the **position** (counting from the upper left corner of the program) and 
2. its **size**. 
In my case I have set up 6 different areas that can be used to position windows; one can use several areas at the same time.

The "change the window **size**" portion works properly. But the "send window to certain location" is buggy (I am quite certain that it's because of Cinnamon side of the execution). Sometimes (most of the times) it works fine, but sometimes the upper left corner of the program is in actuality dozens of pixels off from the place it should be.

I hope I answered your question.

I have used both 1080p and 1440p resolutions.

**Comment by leigh123linux** [comment]
If the apps have a minimum width of 500 it wouldn't be allowed to tile as it would require a screen resolution greater than 1500 width.

https://github.com/linuxmint/cinnamon/issues/8548

**Comment by leigh123linux** [comment]
Run 

```
xprop |grep minimum
```

and it will tell you the minimum window size

**Comment by mikkovedru** [comment]
Firefox: 450 by 95
Nemo: 304 by 163
LibreOffice Writer: 535 by 27

So all of them would fit without any problems in any of the used resolutions ( 1980x1080, let alone 2560x1440; Neither has HiDPI enabled) divided into 3 columns.

And that's how **it works in most cases with 2 out of 3 columns. It's 1 column out of 3 (the left one) that is bugging**.

Evince pdf reader: 689 by 147. So the issue you mentioned might be at play at the 1980x1080 resolution (689*3 = 2067px). But even then, the way it works is that if there is no bug, the **position** (of the upper left corner of the window) should still work properly, while it's the right side of the window that goes beyond the border.

**Comment by mikkovedru** [comment]
This bug still exists in Linux Mint Cinnamon 20.

**Comment by JosephMcc** [comment]
Is there a way to reproduce this without using gTile?

**Comment by mikkovedru** [comment]
The bug is still present with Linux Mint 20.1 Cinnamon (4.8.6).

@JosephMcc
I am not aware. I am quite confident that the bug is seen in other situations and by other people, but it is not necessarily reported because of the seemingly random nature of it and people's inability to figure out steps to reliably reproduce. Installing gTile and changing a few hotkey settings takes only a couple of minutes with reliable results.

**Comment by mikkovedru** [comment]
The bug is still present in Linux Mint 20.2 Cinnamon (5.0.7).

@icarter09 @JosephMcc 
I also found a way to reproduce the bug without using gTile.

## Preparation
1. Have a taskbar on the bottom.
2. Assign hotkeys for the following Window positioning functions (those are the ones I chose, but the actual keys used are irrelevant in this bug):
- Centre window in screen. `Super+R`
- Move window to right edge. `Super+W`
- Move window to top edge. `Super+A`
- Move window to left edge. `Super+Q`
![image](https://user-images.githubusercontent.com/2521942/142783295-792822e7-e2ca-4fe5-8848-f6059bb3acf2.png)

## Bug reproduction 1 - Vertical issues
1. Press `Super+E`. This will open a file manager window.
2. Press `Super+A` to make it go to the top of the screen. You can also move the window manually with your mouse.
3. Use your mouse, grab the lowest edge of the window and drag it all the way down as far as it goes. It should not allow you to go beyond the taskbar.
4. Press `Super+R` to invoke the "center window" command. 
5. BUG: you can now see that the window moved down and now there is an empty space on top, between the top edge of the window and the top edge of the screen itself.

### Possible explanation
What I suspect happens (and where the buggy behavior locates) is that Cinnamon counts incorrectly the center location of the screen and of the window.

I assume that there are 2 ways to count the actual resolution and the center position:
1. by using the full-pixel amount of the screen
2. taking into account the existence of "dead window space" occupied by the taskbar and using only the actually effective area

I suspect that both of those styles are used, but unfortunately in the place in which only one should be used (preferably the second; but the first works as well; the main thing is to use the same one), thus creating is a dissonance (a bug).

This is just a hunch. I haven't actually tested with taskbar being disabled.

## Bug reproduction 2 - Horizontal issues
0. I recommend using some other windows (white browser or file manager windows work well for me) to "mark territory" by aligning edges for easier comparison, otherwise, it might be difficult to see the misalignment of jumping windows, especially on the black background.
1. Press `Super+E`. This will open a file manager window. It's better (easier to align other windows in the background or with partially overlapping vertical edges) if the window doesn't take all the vertical space like in the previous Vertical bug issue. All we care about here is the horizontal proportions/size/edges of the window.
2. Resize the window to take exactly 1/3 of the horizontal space. 
	2.A) You can do it easily with `gTile`.
	2.B) or alternatively you can use the `Super+R` (Centre window in screen) and `Super+W` (Move window to right edge). You can switch between those modes and manually align the edges (right edge of the centered window with the left edge of the window that was moved to the very right edge of the screen). When the edges are aligned it means that mathematically speaking the window should be the size of exactly 1/3 of the screen.
3. BUG: And now switch between `Super+R` (Centre window in screen) and `Super+Q` (Move window to left edge) and see that the edges are not aligning! Sure, you can try to align them again, but after that `Super+R` and `Super+W` will once again show that the windows are not aligned. It happens because the `Super+R` (Centre window in screen) does not actually center the window, neither vertically nor horizontally.


## Comments 
I don't know if fixing the bugs I described in this comment will fix the bug I described in the original issue. But perhaps this will lead to some new ideas about the possible causes of the original bug and all the mentioned bugs will be fixed.

**Comment by mtwebster** [comment]
It looks like the issue is with how gtile handles (or rather, doesn't handle) client-decorated windows like [these](https://user-images.githubusercontent.com/2521942/78409906-9e450100-7613-11ea-99e7-3c0516d5bf85.png). You can identify them by the fact they usually have controls embedded in the titlebar itself.

These windows have frames created by the app itself, not cinnamon/muffin (such as nemo windows). They're pretty confusing to work with because you have to compensate for them in the opposite direction you do for non-client-decorated ones. The gtile extension doesn't take this into consideration (and not surprising, as it's had no real development on it for years).

Try this:
Download:
https://github.com/mtwebster/cinnamon-spices-extensions/raw/fix-gtile-csd-handling/gTile%40shuairan/files/gTile%40shuairan/extension.js

then copy it over this file:
`~/.local/share/cinnamon/extensions/gTile@shuairan/extension.js`

If you're satisfied I'll update the extension permanently.

**Comment by mikkovedru** [comment]
I tried and the original bug is still there even with your updated file. The caveat is that I have not been able to reproduce the bug with the original instructions with Firefox. I also tried to reproduce the bug with several old versions of the extension, but that failed as well. I guess some Firefox/Cinnamon/other developments in the meanwhile changed the bug reproduction situation with Firefox.

But the bug is present beautifully (reproducible every time, no matter which combination of columns I choose) if testing with Evince (pdf reader).

The pixel error I get using gTile is around (precisely?) 2 times the amount of the "center the window" bug I described in my last comment.
![image](https://user-images.githubusercontent.com/2521942/142943709-e2c0baaf-fa6f-49f6-9eb8-15c5cf6352e7.png)

The "center the window" bug I described in my previous comment might give ideas on how to fix the gTile related bug, but doesn't in itself have anything to do with gTile and can't be fixed by improving the gTile extension. Should I create a new issue and copy my last comment in there?
